File: bridge/tools.js
// OpenAI Realtime tool definitions + executors operating on a CallState.
import { findItem, priceOf, hasSizes, sizesOf, generateReceiptNo } from "./menu.js";
import { updateSession, logEvent, createOrder } from "./supabase.js";
import { notifyOrder } from "./whatsapp.js";

export function newCallState({ sessionId, callerMsisdn }) {
  return {
    sessionId,
    callerMsisdn,
    customerName: null,
    cart: [], // [{ id, name, size, qty, unitPence, lineTotal }]
    confirmed: false,
    order: null,
  };
}

function recalc(line) { line.lineTotal = line.unitPence * line.qty; return line; }
function cartTotal(cart) { return cart.reduce((s, l) => s + l.lineTotal, 0); }

async function syncCart(state) {
  await updateSession(state.sessionId, { cart: state.cart });
}

function normaliseSize(size) {
  const s = String(size || "").toLowerCase().trim();
  if (!s) return null;
  if (/^(sml|small)\b/.test(s)) return "Sml";
  if (/^(lrg|large)\b/.test(s)) return "Lrg";
  if (/^(reg|regular)\b/.test(s)) return "Reg";
  return size;
}

function normaliseAddArgs(args = {}) {
  const next = { ...args };
  const rawName = String(next.name || "");
  const q = rawName.toLowerCase().replace(/[^a-z0-9]+/g, " ").trim();
  if (/\b(?:carrot|karrot|kara)\s+chai\b/.test(q)) next.name = "Karak Chai";
  if (/\bsweet\s+spot\s+specials?\b/.test(q) && !/\b(cookie|dough|cheesecake|sundae)\b/.test(q)) {
    next.name = "Sweet Spot Special Waffle";
    if (!next.size) next.size = "Reg";
  }
  next.size = normaliseSize(next.size);
  return next;
}

function normaliseBatchItems(args = {}) {
  const source = Array.isArray(args.items) ? args.items : Array.isArray(args) ? args : [];
  return source
    .map((item) => normaliseAddArgs(item || {}))
    .filter((item) => String(item.name || "").trim());
}

export const TOOL_SCHEMAS = [
  {
    type: "function",
    name: "record_items",
    description: "Add one or more ordered items to the cart in a single call. Default waffles and cookie dough to Reg unless the caller explicitly said small or large. Sweet Spot Special without a category means Sweet Spot Special Waffle Reg. carrot/karrot chai means Karak Chai.",
    parameters: {
      type: "object",
      properties: {
        items: {
          type: "array",
          items: {
            type: "object",
            properties: {
              name: { type: "string" },
              size: { type: "string", description: "Sml, Reg, or Lrg. Use Reg by default for sized desserts." },
              qty: { type: "integer", minimum: 1, default: 1 },
            },
            required: ["name"],
          },
        },
      },
      required: ["items"],
    },
  },
  {
    type: "function",
    name: "add_item",
    description: "Add an item to the order immediately. For waffles and cookie dough, NEVER ask for size first: default to Reg unless the caller explicitly said small or large in this same request. If caller says Sweet Spot Special without a category, add Sweet Spot Special Waffle Reg. If caller says carrot/karrot chai, add Karak Chai.",
    parameters: {
      type: "object",
      properties: {
        name: { type: "string", description: "Item name as the caller said it (e.g. 'kinder waffle', 'oreo shake')." },
        size: { type: "string", description: "Size variant. Use Reg by default for waffles/cookie dough. Use Sml only if caller said small; use Lrg only if caller said large. Omit if N/A." },
        qty: { type: "integer", minimum: 1, default: 1 },
      },
      required: ["name"],
    },
  },
  {
    type: "function",
    name: "remove_item",
    description: "Remove an item from the cart by name (and optional size).",
    parameters: {
      type: "object",
      properties: { name: { type: "string" }, size: { type: "string" } },
      required: ["name"],
    },
  },
  {
    type: "function",
    name: "set_customer_name",
    description: "Record the customer's name when they tell you.",
    parameters: { type: "object", properties: { name: { type: "string" } }, required: ["name"] },
  },
  {
    type: "function",
    name: "read_back_cart",
    description: "Get a structured snapshot of the current cart and total to read back to the caller.",
    parameters: { type: "object", properties: {} },
  },
  {
    type: "function",
    name: "confirm_order",
    description: "Finalize the order, generate a receipt number, persist it, and send WhatsApp confirmations. Only call after the caller has clearly indicated they're done (e.g. 'that's it', 'all done').",
    parameters: { type: "object", properties: {} },
  },
];

export async function execTool(state, name, args) {
  switch (name) {
    case "record_items": {
      const items = normaliseBatchItems(args);
      if (!items.length) return { ok: false, error: "No valid items supplied." };
      const added = [];
      const errors = [];
      for (const itemArgs of items) {
        const result = await execTool(state, "add_item", itemArgs);
        if (result.ok) added.push(result.added);
        else errors.push({ item: itemArgs, error: result.error });
      }
      return { ok: errors.length === 0, added, errors, cart: state.cart, total_pence: cartTotal(state.cart) };
    }

    case "add_item": {
      args = normaliseAddArgs(args);
      const item = findItem(args.name);
      if (!item) {
        await logEvent(state.sessionId, "tool_add_miss", args.name, args);
        return { ok: false, error: `Item "${args.name}" not found on the menu.` };
      }
      let size = args.size || null;
      if (hasSizes(item) && !size) {
        size = "Reg";
      }
      if (size && hasSizes(item) && !item.sizes[size]) {
        return { ok: false, error: `Size "${size}" not available for ${item.base}. Available: ${sizesOf(item).join(", ")}.` };
      }
      const qty = Math.max(1, args.qty || 1);
      const unitPence = priceOf(item, size);
      const line = recalc({ id: item.id, name: item.base, size, qty, unitPence, lineTotal: 0 });
      // Merge if identical line exists
      const existing = state.cart.find((l) => l.id === item.id && l.size === size);
      if (existing) { existing.qty += qty; recalc(existing); }
      else state.cart.push(line);
      await syncCart(state);
      await logEvent(state.sessionId, "cart_add", `${qty}× ${size || ""} ${item.base}`.trim(), { item: item.id, size, qty });
      return { ok: true, added: { name: item.base, size, qty, unitPence }, cart: state.cart, total_pence: cartTotal(state.cart) };
    }

    case "remove_item": {
      const item = findItem(args.name);
      if (!item) return { ok: false, error: `Not on menu: ${args.name}` };
      const before = state.cart.length;
      state.cart = state.cart.filter((l) => !(l.id === item.id && (!args.size || l.size === args.size)));
      await syncCart(state);
      await logEvent(state.sessionId, "cart_remove", item.base, { item: item.id, size: args.size || null });
      return { ok: true, removed: before - state.cart.length, cart: state.cart, total_pence: cartTotal(state.cart) };
    }

    case "set_customer_name": {
      state.customerName = String(args.name || "").trim().slice(0, 80) || null;
      await updateSession(state.sessionId, { customer_name: state.customerName });
      await logEvent(state.sessionId, "name_captured", state.customerName);
      return { ok: true, customer_name: state.customerName };
    }

    case "read_back_cart": {
      return { cart: state.cart, total_pence: cartTotal(state.cart), customer_name: state.customerName };
    }

    case "confirm_order": {
      if (!state.cart.length) return { ok: false, error: "Cart is empty." };
      if (state.confirmed) return { ok: true, already: true, order: state.order };
      const receiptNo = generateReceiptNo();
      const totalPence = cartTotal(state.cart);
      const order = await createOrder({
        sessionId: state.sessionId,
        callerMsisdn: state.callerMsisdn,
        customerName: state.customerName,
        receiptNo,
        cart: state.cart,
        totalPence,
      });
      state.confirmed = true;
      state.order = order;
      await updateSession(state.sessionId, { status: "confirmed", current_intent: "confirmed" });
      await logEvent(state.sessionId, "order_confirmed", receiptNo, { total_pence: totalPence });
      // Fire WhatsApp messages without blocking the voice response
      notifyOrder({ order, callerMsisdn: state.callerMsisdn })
        .then(() => logEvent(state.sessionId, "whatsapp_sent", receiptNo))
        .catch((e) => logEvent(state.sessionId, "whatsapp_error", String(e)));
      return { ok: true, receipt_no: receiptNo, total_pence: totalPence };
    }

    default:
      return { ok: false, error: `Unknown tool ${name}` };
  }
}

File: pipecat/tools.py
"""Tool definitions and executors — Python port of bridge/tools.js.

Uses Pipecat's native FunctionSchema / register_function pattern so tool calls
are reliable across model providers (no more record_items routing bugs).
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Any, Optional

from loguru import logger
from pipecat.adapters.schemas.function_schema import FunctionSchema
from pipecat.services.llm_service import FunctionCallParams

from .menu import find_item, generate_receipt_no, has_sizes, price_of, sizes_of
from .supabase_client import create_order, log_event, update_session


@dataclass
class CallState:
    session_id: str
    caller_msisdn: Optional[str] = None
    customer_name: Optional[str] = None
    cart: list[dict] = field(default_factory=list)
    confirmed: bool = False
    order: Optional[dict] = None


def _recalc(line: dict) -> dict:
    line["lineTotal"] = line["unitPence"] * line["qty"]
    return line


def _cart_total(cart: list[dict]) -> int:
    return sum(l["lineTotal"] for l in cart)


def _normalise_size(size: Optional[str]) -> Optional[str]:
    s = (size or "").lower().strip()
    if not s:
        return None
    if re.match(r"^(sml|small)\b", s):
        return "Sml"
    if re.match(r"^(lrg|large)\b", s):
        return "Lrg"
    if re.match(r"^(reg|regular)\b", s):
        return "Reg"
    return size


def _normalise_add_args(args: dict) -> dict:
    out = dict(args or {})
    raw = str(out.get("name") or "")
    q = re.sub(r"[^a-z0-9]+", " ", raw.lower()).strip()
    if re.search(r"\b(?:carrot|karrot|kara)\s+chai\b", q):
        out["name"] = "Karak Chai"
    if re.search(r"\bsweet\s+spot\s+specials?\b", q) and not re.search(
        r"\b(cookie|dough|cheesecake|sundae)\b", q
    ):
        out["name"] = "Sweet Spot Special Waffle"
        out.setdefault("size", "Reg")
    out["size"] = _normalise_size(out.get("size"))
    return out


# ---------- core operations ----------


def _sync_cart(state: CallState) -> None:
    update_session(state.session_id, {"cart": state.cart})


def _add_item(state: CallState, args: dict) -> dict:
    args = _normalise_add_args(args)
    item = find_item(args.get("name", ""))
    if not item:
        log_event(state.session_id, "tool_add_miss", args.get("name"), args)
        return {"ok": False, "error": f"Item \"{args.get('name')}\" not found on the menu."}
    size = args.get("size")
    if has_sizes(item) and not size:
        size = "Reg"
    if size and has_sizes(item) and size not in item["sizes"]:
        return {
            "ok": False,
            "error": f"Size \"{size}\" not available for {item['base']}. Available: {', '.join(sizes_of(item))}.",
        }
    qty = max(1, int(args.get("qty") or 1))
    unit_pence = price_of(item, size)
    line = _recalc(
        {"id": item["id"], "name": item["base"], "size": size, "qty": qty, "unitPence": unit_pence, "lineTotal": 0}
    )
    existing = next((l for l in state.cart if l["id"] == item["id"] and l["size"] == size), None)
    if existing:
        existing["qty"] += qty
        _recalc(existing)
    else:
        state.cart.append(line)
    _sync_cart(state)
    log_event(
        state.session_id,
        "cart_add",
        f"{qty}x {size or ''} {item['base']}".strip(),
        {"item": item["id"], "size": size, "qty": qty},
    )
    return {
        "ok": True,
        "added": {"name": item["base"], "size": size, "qty": qty, "unitPence": unit_pence},
        "cart": state.cart,
        "total_pence": _cart_total(state.cart),
    }


def _remove_item(state: CallState, args: dict) -> dict:
    item = find_item(args.get("name", ""))
    if not item:
        return {"ok": False, "error": f"Not on menu: {args.get('name')}"}
    before = len(state.cart)
    state.cart = [
        l for l in state.cart if not (l["id"] == item["id"] and (not args.get("size") or l["size"] == args["size"]))
    ]
    _sync_cart(state)
    log_event(state.session_id, "cart_remove", item["base"], {"item": item["id"], "size": args.get("size")})
    return {"ok": True, "removed": before - len(state.cart), "cart": state.cart, "total_pence": _cart_total(state.cart)}


def _confirm_order(state: CallState) -> dict:
    if not state.cart:
        return {"ok": False, "error": "Cart is empty."}
    if state.confirmed:
        return {"ok": True, "already": True, "order": state.order}
    receipt_no = generate_receipt_no()
    total_pence = _cart_total(state.cart)
    order = create_order(
        session_id=state.session_id,
        caller_msisdn=state.caller_msisdn,
        customer_name=state.customer_name,
        receipt_no=receipt_no,
        cart=state.cart,
        total_pence=total_pence,
    )
    state.confirmed = True
    state.order = order
    update_session(state.session_id, {"status": "confirmed", "current_intent": "confirmed"})
    log_event(state.session_id, "order_confirmed", receipt_no, {"total_pence": total_pence})
    # WhatsApp dispatch hook — wire to your sender later (kept non-blocking on purpose).
    return {"ok": True, "receipt_no": receipt_no, "total_pence": total_pence}


# ---------- Pipecat handlers ----------


def build_handlers(state: CallState):
    async def record_items(params: FunctionCallParams) -> None:
        items = params.arguments.get("items") or []
        if not items:
            await params.result_callback({"ok": False, "error": "No valid items supplied."})
            return
        added, errors = [], []
        for raw in items:
            res = _add_item(state, raw)
            if res.get("ok"):
                added.append(res["added"])
            else:
                errors.append({"item": raw, "error": res.get("error")})
        await params.result_callback(
            {
                "ok": not errors,
                "added": added,
                "errors": errors,
                "cart": state.cart,
                "total_pence": _cart_total(state.cart),
            }
        )

    async def add_item(params: FunctionCallParams) -> None:
        await params.result_callback(_add_item(state, params.arguments or {}))

    async def remove_item(params: FunctionCallParams) -> None:
        await params.result_callback(_remove_item(state, params.arguments or {}))

    async def set_customer_name(params: FunctionCallParams) -> None:
        name = str((params.arguments or {}).get("name") or "").strip()[:80] or None
        state.customer_name = name
        update_session(state.session_id, {"customer_name": name})
        log_event(state.session_id, "name_captured", name)
        await params.result_callback({"ok": True, "customer_name": name})

    async def read_back_cart(params: FunctionCallParams) -> None:
        await params.result_callback(
            {"cart": state.cart, "total_pence": _cart_total(state.cart), "customer_name": state.customer_name}
        )

    async def confirm_order(params: FunctionCallParams) -> None:
        try:
            await params.result_callback(_confirm_order(state))
        except Exception as exc:  # noqa: BLE001
            logger.exception("confirm_order failed")
            await params.result_callback({"ok": False, "error": str(exc)})

    return {
        "record_items": record_items,
        "add_item": add_item,
        "remove_item": remove_item,
        "set_customer_name": set_customer_name,
        "read_back_cart": read_back_cart,
        "confirm_order": confirm_order,
    }


TOOL_SCHEMAS: list[FunctionSchema] = [
    FunctionSchema(
        name="record_items",
        description=(
            "Add one or more ordered items to the cart in a single call. Default waffles and "
            "cookie dough to Reg unless the caller explicitly said small or large. Sweet Spot "
            "Special without a category means Sweet Spot Special Waffle Reg. carrot/karrot chai "
            "means Karak Chai."
        ),
        properties={
            "items": {
                "type": "array",
                "items": {
                    "type": "object",
                    "properties": {
                        "name": {"type": "string"},
                        "size": {"type": "string", "description": "Sml, Reg, or Lrg."},
                        "qty": {"type": "integer", "minimum": 1, "default": 1},
                    },
                    "required": ["name"],
                },
            },
        },
        required=["items"],
    ),
    FunctionSchema(
        name="add_item",
        description=(
            "Add a single item to the order. For waffles and cookie dough, default to Reg unless "
            "the caller explicitly said small or large."
        ),
        properties={
            "name": {"type": "string"},
            "size": {"type": "string"},
            "qty": {"type": "integer", "minimum": 1, "default": 1},
        },
        required=["name"],
    ),
    FunctionSchema(
        name="remove_item",
        description="Remove an item from the cart by name (and optional size).",
        properties={"name": {"type": "string"}, "size": {"type": "string"}},
        required=["name"],
    ),
    FunctionSchema(
        name="set_customer_name",
        description="Record the customer's name when they tell you.",
        properties={"name": {"type": "string"}},
        required=["name"],
    ),
    FunctionSchema(
        name="read_back_cart",
        description="Get a snapshot of the current cart and total to read back to the caller.",
        properties={},
        required=[],
    ),
    FunctionSchema(
        name="confirm_order",
        description=(
            "Finalize the order, generate a receipt number, and persist it. Only call after the "
            "caller has clearly indicated they're done."
        ),
        properties={},
        required=[],
    ),
]